### PR TITLE
Add HLL hashing of IP addresses

### PR DESCRIPTION
**Pull request by staticfloat**
We want to persist our logs, but we can't store the raw logs as they contain too much user-identifiable information.  By using HyperLogLog hashing, we can still keep a privacy-preserving value that allows us to (roughly) estimate how many unique IP addresses are represented over a time window, without being able to determine the actual IP addresses themselves (or even the exact number of IP addresses).

This PR adds an `hll_hash` field to the CSVs we're uploading.  It contains a new buildkite secret, which is the encryption key we use to scramble IP addresses while hashing them, which is intended to prevent users from attempting to game our hashing system by providing adversarial spoofed IP addresses.

### CODE REVIEW

File: src/hll_hashing.jl [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+function ip_data(addr::IPAddr)
+    n = sizeof(addr.host)
+    [(addr.host >>> ((n-i) << 3)) % UInt8 for i = 1:n]
+end
+
+ip_addr(host::UInt32) = IPv4(host)
+ip_addr(host::UInt128) = IPv6(host)
+
+# Feistel network with sha256 as round function
```
Comment (StefanKarpinski) [inline src/hll_hashing.jl:16]:
```suggestion
# Feistel network with sha512 as round function
```
I decided to use more bits but didn't fix the comment to match.

File: src/hll_hashing.jl [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    data = Base.StringVector(3)
+    data[1] = bucket % UInt8
+    data[2] = bucket >> 8
+    data[3] = sample + 1
+    return String(data)
+end
+
+# By default, we encrypt with a zeroed-out key.  The user must either explicitly pass in `key` to `hll_hash_ip()`
+# or call `load_hll_key!()` to change this key value.
+const default_hll_key = zeros(UInt8, 512)
```
Comment (StefanKarpinski) [inline src/hll_hashing.jl:74]:
```suggestion
const default_hll_key = UInt8[]
```
The key can actually be any length, so we could just use an empty key here.

File: src/hll_hashing.jl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+function load_hll_key!(path::AbstractString)
+    open(path, read=true) do io
+        # Read the file, try to assign it into `default_hll_key`
+        new_key = read(io)
+        if length(new_key) != 512
```
Comment (StefanKarpinski) [inline src/hll_hashing.jl:80]: The key can be any length — see comment above.

File: src/hll_hashing.jl [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+const DIGEST = SHA.SHA512_CTX
+
+function feistel!(key::Vector{UInt8}, data::Vector{UInt8}, enc::Bool)
+    n = length(data)
+    m = n >> 1
+    rounds = enc ? (0:1:ROUNDS-1) : (ROUNDS-1:-1:0)
+    for round in rounds
+        digest = DIGEST()
+        update!(digest, [round % UInt8])
+        update!(digest, key)
```
Comment (StefanKarpinski) [inline src/hll_hashing.jl:28]: This is the only use of the key and it doesn't care about how much data is in the key. Its only purpose is to make the Feistel round function hard to predict/invert. In some sense it is choosing a permutation of all possible data values. I.e. for 32-bit data (4 byte), it is choosing one of the `(2^32)! ≈ 2^2.6e35` permutations of all the possible 32-bit values. Of course, we can't possibly use a key large enough to access all of those possible permutations — we'd need a key more than 2.6e35 bits long. But using a key with N bits, for N = 512, 1024 or whatever, is pretty good since it chooses one of 2^N possible permutations of 32-bit values at random and the complex, uninvertible nature of the SHA512 function means that it's basically impossible to understand which permutations are the possible set of permutations, which is good enough.

TLDR: the key can be any data at all, no fixed length.
